internal/command/action_v2_target_key.go

```diff
@@ -0,0 +1,187 @@
+package command
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	"github.com/zitadel/zitadel/internal/repository/target"
+	"github.com/zitadel/zitadel/internal/telemetry/tracing"
+	"github.com/zitadel/zitadel/internal/zerrors"
+)
+
+const (
+	ErrTargetIDMissing          = "ErrTargetIDMissing"
+	ErrPublicKeyFormat          = "ErrPublicKeyFormat"
+	ErrExpirationDateBeforeNow  = "ErrExpirationDateBeforeNow"
+	ErrPublicKeyDeleteActiveKey = "ErrPublicKeyDeleteActiveKey"
+)
+
+type TargetPublicKey struct {
+	TargetID   string
+	PublicKey  []byte
+	Expiration time.Time
+
+	// KeyID is generated and returned after adding the public key
+	KeyID string
+}
+
+func (c *Commands) AddTargetPublicKey(ctx context.Context, key *TargetPublicKey, resourceOwner string) (_ time.Time, err error) {
+	ctx, span := tracing.NewSpan(ctx)
+	defer func() { span.EndWithError(err) }()
+
+	if key.TargetID == "" {
+		return time.Time{}, zerrors.ThrowInvalidArgument(nil, ErrTargetIDMissing, "Errors.IDMissing")
+	}
+	if !key.Expiration.IsZero() && key.Expiration.Before(time.Now()) {
```

Comment:
I believe that zero `Time.time` is always before `time.Now()`, so checking for `key.Expiration.Before(time.Now())` should cover also for the `IsZero()` check